src/matrixvariate.jl

```diff
@@ -124,45 +124,56 @@ end
 
 #### Sampling
 function Distributions._rand!(rng::AbstractRNG, d::TuringWishart, A::AbstractMatrix)
-    _wishart_genA!(rng, Distributions.dim(d), d.df, A)
+    Distributions._wishart_genA!(rng, Distributions.dim(d), d.df, A)
     unwhiten!(d.chol, A)
     A .= A * A'
 end
 
-function _wishart_genA!(rng::AbstractRNG, p::Int, df::Real, A::AbstractMatrix)
-    # Generate the matrix A in the Bartlett decomposition
-    #
-    #   A is a lower triangular matrix, with
-    #
-    #       A(i, j) ~ sqrt of Chisq(df - i + 1) when i == j
-    #               ~ Normal()                  when i > j
-    #
-    A .= zero(eltype(A))
-    for i = 1:p
-        @inbounds A[i,i] = rand(rng, Chi(df - i + 1.0))
-    end
-    for j in 1:p-1, i in j+1:p
-        @inbounds A[i,j] = randn(rng)
-    end
-end
-
 function unwhiten!(C::Cholesky, x::StridedVecOrMat)
     cf = C.U
     lmul!(transpose(cf), x)
 end
 
+## Custom adjoint since Zygote can't differentiate through `@warn`
+
+ZygoteRules.@adjoint function Wishart(df::T, S::AbstractPDMat{T}, warn::Bool = true) where T
+    return ZygoteRules.pullback(_Wishart, df, S, warn)
+end
+
+function _Wishart(df::T, S::AbstractPDMat{T}, warn::Bool = true) where T
+    df > 0 || throw(ArgumentError("df must be positive. got $(df)."))
+    p = dim(S)
+    rnk = p
+    singular = df <= p - 1
+    if singular
+        isinteger(df) || throw(ArgumentError("singular df must be an integer. got $(df)."))
+        rnk = convert(Integer, df)
+        warn && _warn("got df <= dim - 1; returning a singular Wishart")
+    end
+    logc0 = Distributions.wishart_logc0(df, S, rnk)
+    R = Base.promote_eltype(T, logc0)
+    prom_S = convert(AbstractArray{T}, S)
+    Wishart{R, typeof(prom_S), typeof(rnk)}(R(df), prom_S, R(logc0), rnk, singular)
+end
+
+_warn(msg) = @warn(msg)
+
+ZygoteRules.@adjoint function _warn(msg)
+    return _warn(msg), _ -> nothing
+end
```

Comment:
Is there a better solution than just copying the constructor?

Reply:
I don't think so.